finetune/base.py

```diff
@@ -92,34 +91,73 @@ def cleanup():
+            if self.config.float_16_predict or self.config.mixed_precision:
+                LOGGER.warning(
+                    "This GPU does not support float 16 ops but they were requested in the config. They are being turned off."
+                )
+                self.config = self.resolve_config(no_fp16=True)
+
+    def resolve_config(self, no_fp16=False, **kwargs):
+        self.config_overrides.update(kwargs)
+        assert_valid_config(**self.config_overrides)
         config = get_default_config()
         config.base_model = kwargs.get("base_model", config.base_model)
 
         auto_keys = []
-        config.update(kwargs)
+        config.update(self.config_overrides)
         for k, v in config.items():
             if isinstance(v, str) and v.lower() == "auto":
                 auto_keys.append(k)
-            elif k in kwargs:
+            elif k in self.config_overrides:
                 continue
             elif k in config.base_model.settings:
                 config[k] = config.base_model.settings[k]
 
-        overrides = config.base_model.get_optimal_params(config)
-
-        if not issubclass(config.base_model, _BaseBert) and config.float_16_predict:
-            LOGGER.warning("float_16_predict only supported by bert based models")
+        # This has to be here before optimal_params are derived because some are dependant on these values.
+        if (not issubclass(config.base_model, _BaseBert) or no_fp16) and (
+            config.float_16_predict == True or config.mixed_precision == True
+        ):
+            LOGGER.warning(
+                "float_16_predict and mixed_precision only supported by bert based models and Volta or newer GPUs"
+            )
             config.float_16_predict = False
+            config.mixed_precision = False
+
+        if (
+            not issubclass(config.base_model, _BaseBert) or no_fp16
+        ) and "fp16" in config.optimize_for:
+            new_optimize_for = config.optimize_for.replace("_fp16", "")
+            LOGGER.warning(
+                "optimize_for was set to {} but fp16 is not supported by this gpu so falling back to {}".format(
+                    config.optimize_for, new_optimize_for
+                )
+            )
+            config.optimize_for = new_optimize_for
 
+        overrides = config.base_model.get_optimal_params(config)
         for ak in auto_keys:
             if ak in ["val_size", "use_gpu_crf_predict"]:
                 continue  # this auto is resolved after data is provided.
             if ak not in overrides:
                 raise ValueError("There is no auto setting for {}".format(ak))
             config[ak] = overrides[ak]
+
+        if hasattr(self, "input_pipeline"):
+            self.input_pipeline.config = config
+
+        if hasattr(self, "config"):
+            for k, v in self.config.items():
+                if k not in auto_keys and k != "optimize_for" and config[k] != v:
+                    LOGGER.warning(
```

Comment:
Wouldn't this end up being quite noisy?  I would expect this to be virtually always hit.